### ISSUE TITLE
Question: How really does memory leak detection work...

### ISSUE THREAD

**Issue by arstrube** [cpputest/cpputest#533]
...in conjunction with the MemoryLeakDetectorXXXMacros.h?

**Comment by arstrube** [comment]
I am confused.

Originally, I thought they won't work without the headers. Period.

That doesn't seem to be the case, though. Or not quite.

As I am testing the thread-safe versions of the memory management overloads, I find that when CppUTest is compiled without flags and without the abovementioned headers, the new / delete macros will still work as expected.

However, the malloc / free memory leak detection does not seem to be activated at all.

**Comment by arstrube** [comment]
I mean, the function pointers are activated, but there doesn't seem anything that makes malloc() or free() use them?

**Comment by Andne** [comment]
I noticed the same behavior in one of my projects last week.  Judging from the contents of the forced include files, cpputest doesn't actually implement the malloc and free functions, it turns them into macros that call a different set of functions.  This means that unless `MemoryLeakDetectorMallocMacros.h` is included in a file, it won't see the overrides.  Whereas the new and delete functions (operators) do get implemented, so those get used even if the forced include is not in place (assuming that linker decides it wants to).  The one thing you do lose by not having `MemoryLeakDetectorNewMacros.h` included are a couple of overrides that add the `__FILE__` and `__LINE__` macros into the call, so reported leaks are harder to debug.

**Comment by arstrube** [comment]
Thanks Andy. yeah, that explains a lot.

I rewrote the test to call cpputest_malloc() & friends directly (for my current test, that doesn't matter).

You lose more than you might think, though... For example the possibility to use the threadsafe versions of these functions. That's what happened in the test -- they weren't called, so the stub never incremented the mutex count, so the my test failed.

In fact, if I interpret matters correctly, you lose the possibility to have memory leaks reported entirely, because 'original' malloc() & friends will be called directly, as far as I can tell. (@basvodde - correct me if I am wrong about this).

**Comment by Andne** [comment]
Guess I didn't say that very well, let me try again.
- malloc and free need the forced include in order for any leaks to be detected.
- new and delete will detect leaks without the forced include, but they won't have the location details unless the forced include was used.

**Comment by arstrube** [comment]
Roger that :-)

**Comment by basvodde** [comment]
Close?

**Comment by arstrube** [comment]
Close. Question has been answered.

**Comment by arstrube** [comment]
P.S. If anyone reads this, please also note that the `new` overloads will be silently ignored when compiled with Clang and -02 option. Beware!
